### PR TITLE
WIP: ENH: Integrating LAPACK "expert" routines with conditioning warnings for linalg.solve_ family

### PR DESCRIPTION
**Pull request by ilayn**
As briefly discussed in #6634, in this PR the ?GESVX routine of LAPACK is exposed for generic use. They are the so-called eXpert versions of the already existing ?GESV routine family which are used in `sp.linalg.solve` 

The use case for these routines are many-fold 
- The most important case is to finally change the silent singular matrix handling. With the condition number and error estimates, a warning/error/msg can be emitted (needs-decision).
- Ill-conditioned matrices can be checked before inversion steps and nonsensical results can be avoided
- cond and rcond functions can now use estimates for comparison
- the lu factorization comes with a bonus (as is with ?GESV too but the API throws it away currently)
- The numerical accuracy is improved thanks to the balancing and iterative refinement (Lapack prefers the term `equilibration` for this particular instance). 

The API requirements for such routine is a bit tedious to satisfy. One of the main reasons for that is the variety of options to choose from. To give an example, here is an example from LAPACK documentation: 

> ```
>      EQUED is CHARACTER*1
>       Specifies the form of equilibration that was done.
>       = 'N':  No equilibration (always true if FACT = 'N').
>       = 'R':  Row equilibration, i.e., A has been premultiplied by
>               diag(R).
>       = 'C':  Column equilibration, i.e., A has been postmultiplied
>               by diag(C).
>       = 'B':  Both row and column equilibration, i.e., A has been
>               replaced by diag(R) * A * diag(C).
>       EQUED is an input argument if FACT = 'F'; otherwise, it is an
>       output argument.
> ```

or an even more joyful one is the following 

> ```
>       AF is DOUBLE PRECISION array, dimension (LDAF,N)
>       If FACT = 'F', then AF is an input argument and on entry
>       contains the factors L and U from the factorization
>       A = P*L*U as computed by DGETRF.  If EQUED .ne. 'N', then
>       AF is the factored form of the equilibrated matrix A.
> 
>      If FACT = 'N', then AF is an output argument and on exit
>       returns the factors L and U from the factorization A = P*L*U
>       of the original matrix A.
> 
>      If FACT = 'E', then AF is an output argument and on exit
>       returns the factors L and U from the factorization A = P*L*U
>       of the equilibrated matrix A (see the description of A for
>       the form of the equilibrated matrix).
> ```

There will be a a few phases for this PR to come to fruition and I would really appreciate all the feedback from the wizards.
- [x] Make sure the f2py interface is correct. 
- [x] Make sure that the input-output arguments are handled correctly.
- [x] Make sure that the varying number of I/O arguments are not exposed to the user and the decision tree should not lead to confusion. In other words, if out of inexperience a user selects two conflicting keyword arguments (as it is extremely easy to fall into those traps) the behavior should be unambiguous override or err out directly. 
- [ ] Make sure that the tests cover sufficient examples
- [x] Decide on the numerical bound formula to accept numerical singularity
- [ ]  Decide on where this functionality goes into, existing `sp.linalg.solve` as an addition or a standalone solver and so on.

### CODE REVIEW

**Comment by ilayn** [comment]
Can someone help me with this build error ? I can't see what the culprit is.

**Comment by Eric89GXL** [comment]
Does it build for you locally?

I can't help much, but I can say that having a setup where I could iterate locally allowed me to try various fixes much more quickly than waiting for Travis to get it done.

**Comment by ilayn** [comment]
Unfortunately, I'm  stuck with a Windows machine lately hence I can't build SciPy. Even if I had Linux machine I've had really unsuccesful runs in the past with all the compiler flags and settings. 

I was wondering if anyone can recognize the error because as I search more and more it looks like a config-related error from the look of it unless I haven't blinded myself.

**Comment by person142** [comment]
> Unfortunately, I'm stuck with a Windows machine lately hence I can't build SciPy.

Admittedly I've never tried this on a Windows machine, but maybe use VirtualBox + Linux?

> Even if I had Linux machine I've had really unsuccesful runs in the past with all the compiler flags and settings.

I can't attest to how things were in the past, but these days building on say Ubuntu doesn't require any fiddling--just apt-getting some stuff and then running `setup.py`.

Just to second what @Eric89GXL said, I really think jumping through some hoops to get a good dev setup will save you pain in the long run.

**Comment by ilayn** [comment]
I'm doing that as we speak ;) I've asked for admin rights on my work machine so that I can install Linux as a dualboot option.

**Comment by pv** [comment]
You can install Linux also on a virtual machine, for example [Virtualbox](https://www.virtualbox.org/wiki/Downloads) is free. For compilation on Windows, you can try [mingwpy](https://mingwpy.github.io/mingwpy.html) --- never used it though, so can't help if problems appear.

File: scipy/linalg/flapack.pyf.src [hunk 1]
```diff
@@ -542,6 +542,40 @@ interface
...
+
+     <ftype> dimension(n,n),check(shape(a,0)==shape(a,1)),intent(in,out,copy,out=as):: a
+     integer depend(a),intent(hide):: lda = shape(a,0)
+     <ftype> dimension(n,nrhs),intent(in,out,copy,out=bs):: b
+     integer depend(b),intent(hide):: ldb = shape(b,0)
+     integer depend(a),intent(hide):: n = shape(a,0)
+     integer depend(b),intent(hide):: nrhs = shape(b,1)
+     <ftype> dimension(n,nrhs),depend(n,nrhs),intent(out):: x
+     integer depend(n),intent(hide):: ldx = n
+     character intent(in),check(*trans=='N'||*trans=='T'||*trans=='C'):: trans
```
Comment (ev-br) [inline scipy/linalg/flapack.pyf.src]: I'm not sure this is correct --- how exactly f2py works is a bit of a mystery, so it might very well not be the culprit --- but still.
`gemm` does it differently at least (at *that* was written by the original author of f2py)
https://github.com/scipy/scipy/blob/master/scipy/linalg/fblas_l3.pyf.src#L23

File: scipy/linalg/flapack.pyf.src [hunk 2]
```diff
@@ -542,6 +542,40 @@ interface
...
+     character intent(in),check(*fact=='F'||*fact=='N'||*fact=='E'):: fact
+     <ftype> dimension(n,n),intent(in,out,out=lu):: af
```
Comment (ev-br) [inline scipy/linalg/flapack.pyf.src]: Might need a `depend(n)`.

**Comment by ev-br** [comment]
One last comment from the peanut gallery: note that gesvx routines are exposed via cython_lapack.

File: scipy/linalg/flapack.pyf.src (hunk 1 above)
Comment (ilayn) [inline scipy/linalg/flapack.pyf.src]: I would rather use "black magic" but yes it is indeed mysterious. i haven't committed yet but I finally managed the get syntax to wrap some function without any errors that compiles. Now it is "only" causing segmentation faults :) so one step at a time I guess. I'm still fighting with Windows build chain it is really annoying. But anyways.

**Comment by ilayn** [comment]
Ouch, does that mean we don't even need to use f2py? 

Or did you mean as an alternative mean to test the f2py version?

**Comment by ilayn** [comment]
I am having difficulties about choosing the right API for the frontend for this solver. GESVX has too many options for the regular user but the options are quite useful. I have laid the general options available for the interested for investigation but I am not sure about which direction it should evolve towards. 

The main issue is the `fact='F'` option. This is mainly for repeated calls for the solver. If `A` matrix is the same but the right hand side is changing, there is an option to provide the `P`,`L`,`U` factors such that this factorization can be skipped in the next calls. Moreover, there is a equilibration step very much like balancing via `equed` keyword and these scalings `r, `c` should be also supplied, should the user choose to provide these arguments manually. 

Should we skip those and let the experienced users call `gesvx` via `scipy.linalg.lapack.?gesvx` or should we lay bare all the available options ?

File: scipy/linalg/basic.py [hunk 3]
```diff
@@ -105,6 +105,92 @@ def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
...
+          'N'      a   x = b      None
+          'T'      a^T x = b   Transpose
+          'C'      a^H x = b  Complex Conj
+        =========  ========= ==============
+    fact : {'E', 'N', 'F'}, optional
+        Selects the type of pre-conditioning on the data matrices. Specifies
+        whether or not the factorized form of the matrix ``a`` is supplied on
+        entry, and if not, whether the matrix ``a`` should be equilibrated
+        before it is factorized. See Notes for more details.
+    equed : {'B', 'N', 'R', 'C'}, optional
```
Comment (rgommers) [inline scipy/linalg/basic.py]: better to use understandable variable names, not whatever LAPACK calls it. So something like ``equilibrate``. And ``factor`` or ``factorize`` above.

**Comment by rgommers** [comment]
That's indeed a lot of parameters. And return values - some of those also don't seem super useful. I'd suggest to select a sensible subset. If you provide it all, what is the difference between this function and direct use of the underlying routine?

**Comment by rgommers** [comment]
> Should we also bother to mimic such automatic behavior? 

Looking at the API and the problem that ``gesvx`` solves, I'm wondering if this should really be a separate function - the overlap with ``solve`` is very large. Can you expose or switch to ``gesvx`` within ``solve`` you think?

**Comment by ilayn** [comment]
@rgommers I agree with you. That's what I tried to formulate in #6634 last comment. The reason why I did this as a separate funtion is to make the review a little bit independent from the clutter and to give a chance to compare with the old version. It is not meant to be a separate function but rather a prototype. 

The utopia is we embed all solvers to a single frontend and have linsolve like switching behavior but I'll wait for your feedback about performance and backwards compatibility concerns.

**Comment by person142** [comment]
> The utopia is we embed all solvers to a single frontend and have linsolve like switching behavior

Notions of utopia differ. I do not like the idea because I think users should think carefully about which solver they want to use and then use that one. Essentially I think this becomes a black box, and I don't like black boxes. This is just one opinion though, and I'm not confident at all that anyone else will share it, so don't take it too seriously. ;-)

**Comment by ilayn** [comment]
Utopia in the sense that `A\b` is highly optimized and does the choice automatically but if you have a specific matrix you use `linsolve (A,b,'lt')` type of switch to bypass in matlab. 

The burning issue is the silent pass of very ill conditioned solutions and missing LinAlgError raises due to numerical noise. But while we are at it maybe a design change of the API can help users.

**Comment by rgommers** [comment]
> The reason why I did this as a separate funtion is to make the review a little bit independent

That's a good idea. Changing the PR title to WIP to indicate it shouldn't be merged as is.

**Comment by ilayn** [comment]
This is the basic structure that I have in mind for the "expert" driver usage. Performance-wise the expert drivers are roughly 2-5 times slower on average compared to their regular relatives (only tested up to 1000x1000). However, as far as i have tested the warnings are functional (will add tests). 

This might be a disadvantage within the SciPy solve mechanism but sp.linalg.solve is already slow compared to NumPy version probably due to gufunc usage. I don't know much about that machinery but having warnings and better numerical singularity catch in my opinion better than speed.

Also there are options for the positive definite matrix (`sym_pos` and `lower` keys)  from the early days and I think it is cluttering the signature. Instead I've replaced it with a keyword argument which selects from a richer range. I might add triangular, banded matrices and add more but since this is still a prototype, I decided to stop at these options. 

Thus, I need a decision from the involved parties whether this regression on the performance is acceptable or not.

In short, there will be two points regarding backwards compatibility: The two keywords `sym_pos`, `lower` and also the slow-down.

**Comment by person142** [comment]
Isn't `lower` still relevant? Otherwise how do you know which triangle to ignore?

**Comment by ilayn** [comment]
Yes, in the generic case it is not but for the positive definite case it is still relevant. If we decide to subsume the triangular cases into this it will also be relevant too. 

I was a bit reluctant to create specific keywords for each solver in order to avoid clashing/confusion but it looks like we don't have too many specific keywords and the users can set them without getting confused too much. 

I want to place a test with catching a warning, do we have a specific way of working for such things? The idea is to make sure that ill-conditioned matrices emit warnings.

**Comment by ilayn** [comment]
What I would like to consult you guys is whether this is somewhat along the correct direction. Otherwise we can just leave the f2py updates and people can write their own solvers using `get_lapack_funcs`. 

I'm also very interested in what NumPy guys are doing with ufuncs but I don't have any knowledge of those kind of manipulations. (frankly, getting f2py part working was crazy enough 😃) It looks so juicy though in terms of speed.

**Comment by person142** [comment]
> I want to place a test with catching a warning, do we have a specific way of working for such things?

```python
import warnings
from numpy.testing import assert_raises

with warnings.catch_warnings():
    warnings.simplefilter('error')
    assert_raises(...)
```

**Comment by ilayn** [comment]
ah ok, so I can use the standard way. Cool, thanks.

**Comment by ilayn** [comment]
Here are the speed comparison calling directly `sp.linalg.lapack.dgesvx(A,B)` and `sp.linalg.lapack.dgesv(A,B)` with 
```
A = np.random.rand(n, n)
B = np.random.rand(n, 3)
```

horizontal axis is `n` and vertical axis is seconds in the first one. In the second it's just the ratio of two curves. The numbers are averages of 25 runs for each n.

**Comment by ilayn** [comment]
Due to github screw-ups on my fork I'll reopen another PR with a clean slate.
